usr/src/uts/common/fs/zfs/dmu_send.c

```diff
@@ -2644,6 +2644,17 @@ receive_read_record(struct receive_arg *ra)
 		uint32_t size = P2ROUNDUP(drro->drr_bonuslen, 8);
 		void *buf = kmem_zalloc(size, KM_SLEEP);
 		dmu_object_info_t doi;
+
+		if (drro->drr_compress >= ZIO_COMPRESS_FUNCTIONS)
+			return (SET_ERROR(ENOTSUP));
```

Comment:
It looks like you need to `kmem_free(buf, size);` before returning (or better, allocate it after these checks).